net/sync_dag.go

```diff
@@ -58,37 +48,56 @@ func syncDAG(ctx context.Context, bserv blockservice.BlockService, block *corebl
 		return err
 	}
 
-	ssb := builder.NewSelectorSpecBuilder(basicnode.Prototype.Any)
-	matchAllSelector, err := ssb.ExploreRecursive(selector.RecursionLimitNone(), ssb.ExploreUnion(
-		ssb.Matcher(),
-		ssb.ExploreAll(ssb.ExploreRecursiveEdge()),
-	)).Selector()
+	err = loadBlockLinks(ctx, lsys, block)
 	if err != nil {
 		return err
 	}
+	return nil
+}
 
-	// prototypeChooser returns the node prototype to use when traversing
-	prototypeChooser := func(lnk ipld.Link, lnkCtx ipld.LinkContext) (ipld.NodePrototype, error) {
-		if tlnkNd, ok := lnkCtx.LinkNode.(schema.TypedLinkNode); ok {
-			return tlnkNd.LinkTargetNodePrototype(), nil
-		}
-		return basicnode.Prototype.Any, nil
-	}
-	// preloader is used to asynchronously load blocks before traversing
-	//
-	// any errors encountered during preload are ignored
-	preloader := func(pctx preload.PreloadContext, l preload.Link) {
-		go lsys.Load(linking.LinkContext{Ctx: pctx.Ctx}, l.Link, coreblock.SchemaPrototype) //nolint:errcheck
-	}
-	config := traversal.Config{
-		Ctx:                            ctx,
-		LinkSystem:                     lsys,
-		LinkVisitOnlyOnce:              true,
-		LinkTargetNodePrototypeChooser: prototypeChooser,
-		Preloader:                      preloader,
+// loadBlockLinks loads the links of a block recursively.
+//
+// If it encounters errors in the concurrent loading of links, it will return
+// the first error it encountered.
+func loadBlockLinks(ctx context.Context, lsys linking.LinkSystem, block *coreblock.Block) error {
+	ctx, cancel := context.WithTimeout(ctx, syncDAGTimeout)
+	defer cancel()
+
+	var wg sync.WaitGroup
+	var asyncErr error
+	var asyncErrOnce sync.Once
+
+	setAsyncErr := func(err error) {
+		asyncErr = err
+		cancel()
 	}
-	visit := func(p traversal.Progress, n datamodel.Node) error {
-		return nil
+
+	for _, lnk := range block.Links {
+		wg.Add(1)
+		go func(lnk coreblock.DAGLink) {
+			defer wg.Done()
+			if ctx.Err() != nil {
+				return
+			}
+			nd, err := lsys.Load(linking.LinkContext{Ctx: ctx}, lnk, coreblock.SchemaPrototype)
```

Comment:
question: with this change doesn't it mean we can read directly from `datastore.Blockstore` in `mergeProcessor.loadComposites` instead of pulling it from block service?

Reply:
`loadComposites` doesn't use a block service. It already reads directly from the blockstore.

Reply:
alright, I see